File: wind-composer/tempo_engine.py
"""Phase 6 tempo smoothing — max 1 BPM per bar, gust boost decay."""

from __future__ import annotations

from utils import clamp


class TempoEngine:
    def __init__(self) -> None:
        self._current_bpm = 100.0
        self._gust_boost = 0.0
        self._last_bar = -1

    @property
    def current_bpm(self) -> float:
        return self._current_bpm

    def reset(self, bpm: float) -> None:
        self._current_bpm = bpm
        self._gust_boost = 0.0
        self._last_bar = -1

    def update(
        self,
        target_bpm: float,
        bpm_min: float,
        bpm_max: float,
        measure: int,
        gust: bool,
    ) -> float:
        clamped_target = clamp(target_bpm, bpm_min, bpm_max)
        if gust:
            self._gust_boost = min(2.0, self._gust_boost + 2.0)

        if measure != self._last_bar and measure >= 0:
            self._last_bar = measure
            goal = clamp(clamped_target + self._gust_boost, bpm_min, bpm_max + 2.0)
            diff = goal - self._current_bpm
            if abs(diff) > 0.001:
                self._current_bpm += clamp(diff, -1.0, 1.0)
            self._gust_boost *= 0.55
            if self._gust_boost < 0.05:
                self._gust_boost = 0.0

        return clamp(self._current_bpm, bpm_min, bpm_max + 2.0)


def wind_to_target_bpm(
    wind_kmh: float,
    bpm_min: float,
    bpm_max: float,
    trend_wind_delta: float = 0.0,
    storm_likelihood: float = 0.0,
) -> float:
    trend_wind = wind_kmh + trend_wind_delta * 0.35
    factor = clamp(trend_wind / 45.0)
    return bpm_min + factor * (bpm_max - bpm_min) + storm_likelihood * 4.0

File: wind-composer-web/src/music/tempoEngine.ts
import { clamp } from "../utils";

/** Smooths BPM: max 1 BPM change per bar; gust adds temporary +2 BPM that decays. */
export class TempoEngine {
  private currentBpm = 100;
  private gustBoost = 0;
  private lastBar = -1;

  getBpm(): number {
    return this.currentBpm;
  }

  reset(bpm: number): void {
    this.currentBpm = bpm;
    this.gustBoost = 0;
    this.lastBar = -1;
  }

  update(
    targetBpm: number,
    bpmMin: number,
    bpmMax: number,
    measure: number,
    gust: boolean,
  ): number {
    const clampedTarget = clamp(targetBpm, bpmMin, bpmMax);
    if (gust) this.gustBoost = Math.min(2, this.gustBoost + 2);

    if (measure !== this.lastBar && measure >= 0) {
      this.lastBar = measure;
      const goal = clamp(clampedTarget + this.gustBoost, bpmMin, bpmMax + 2);
      const diff = goal - this.currentBpm;
      if (Math.abs(diff) > 0.001) {
        this.currentBpm += clamp(diff, -1, 1);
      }
      this.gustBoost *= 0.55;
      if (this.gustBoost < 0.05) this.gustBoost = 0;
    }

    return clamp(this.currentBpm, bpmMin, bpmMax + 2);
  }
}

export function windToTargetBpm(
  windKmh: number,
  bpmMin: number,
  bpmMax: number,
  trendWindDelta = 0,
  stormLikelihood = 0,
): number {
  const trendWind = windKmh + trendWindDelta * 0.35;
  const factor = clamp(trendWind / 45);
  return bpmMin + factor * (bpmMax - bpmMin) + stormLikelihood * 4;
}
